### ISSUE TITLE
Need way of passing command to override config, in this case for changing varnish config params

### ISSUE THREAD

**Issue by chrisjosephs** [Azure/app-service-quickstart-docker-images#264]
> Please provide us with the following information:
> ---------------------------------------------------------------

### This issue is for a: (mark with an `x`)

- [ ] bug report -> please search issues before submitting
- [x ] feature request
- [ ] documentation issue or request
- [ ] regression (a behavior that used to work and stopped in a new release)
drupal-nginx-fpm/0.51

### Minimal steps to reproduce
> No way of changing varnish header size.Varnish has a max header size of 8kb.  If too many blocks are added then this can crash varnish by creating a X varnish block cache header > 8kb.  If this  many blocks really are needed then http_resp_hdr_len and http_req_hdr_len needs to be boosted in varnish.  However there is no way of setting this automatically on container start.

### Any log messages given by the failure
>

### Expected/desired behavior
Version - drupal-nginx-fpm/0.51
> Should be able to change configuration options by for example passing in a "command" from docker-compose.yml, however currently the provided entrypoint.sh does not wait for commands otherwise could pass in " command: varnishadm param.set http_resp_hdr_len 16kb" etc as well as to override other container settings.  

> ---------------------------------------------------------------
> Thanks! We'll be in touch soon.

**Comment by JCFerrerG** [comment]
Which image does this request pertain to?

For my projects, I will sometimes override the entire config file by mounting a copy of the file as a volume from docker-compose.yml.

**Comment by LeonZhang77** [comment]
@JCFerrerG  you  are a high-end player...so glad to see you. ..

@chrisjosephs yes, 0.51 isn't allow user to change config of php/nginx/varnish when depoly it to azure, Everything you modified will be lost after restart. i am thinking about this, please waiting next version.

**Comment by JCFerrerG** [comment]
Below are some ideas for the purposes of brainstorming.  

Notice that most of these techniques can also be used to override PHP, Nginx, and Varnish settings -- although see #265 for issues with overriding nginx configs.

# Possible solutions to override php/nginx/varnish configs when deploying to Azure

1. ~**Override config files using volumes**~

    This example assumes that you have a custom default.vcl file that is accessible to the Docker host.  
    
    **This may not be feasible on Azure App Service for Containers.** @LeonZhang77 , do know if this technique is feasible on App Service for Containers?

    * From docker CLI

      ```bash
      docker run -v $(pwd)/default.vcl:/etc/default/default.vcl drupal-nginx-fpm:0.51
      ```

    * From docker-compose.yml
     ```yaml
      version: '3'
      services:
          drupal:
              image: drupal-nginx-fpm:0.51
          volumes:
              - ./default.vcl:/etc/default/default.vcl
     ```


2. **Create your own Docker image with custom configs**

   I do this when option 1 is not feasible, as this solution introduces some more work to manage your own images.  However, [Azure Container Registry](https://azure.microsoft.com/en-us/services/container-registry/) makes this very simple to deploy.

    ```Dockerfile
    # Dockerfile for your custom image
    FROM drupal-nginx-fpm:0.51
    
    COPY default.vcl /etc/varnish/default.vcl  # default.vcl is your custom config file
    
    # NOTE: Other config files are easily overridden in the same manner
    #COPY nginx.conf /etc/nginx/
    #COPY php.ini $PHP_CONF_FILE
    ```

3. **Create environment variables for specific services (e.g. VARNISH_OPTS)**

    I have seen a few examples of this in other Dockerfiles. I believe this would require modification of the entrypoint.sh, like in the [Easycom-Solution/docker-varnish example](https://github.com/Easycom-Solutions/docker-varnish/blob/master/docker-entrypoint.sh).  More examples:

    * https://github.com/Easycom-Solutions/docker-varnish/blob/master/docker-entrypoint.sh
    * https://info.varnish-software.com/blog/varnish-docker
        * https://github.com/gquintard/docker-howto/blob/master/bundle/root/varnish-suite.sh
    * https://community.bitnami.com/t/changing-varnish-cache-configuration/44681/2

    My biggest concern about this approach is that it doesn't extend to other services -- like PHP and Nginx -- without more modifications to this image.  This trains the users of the image to be dependent on built-in features -- which could possibly lead to a complex and confusing container definition.

4. ~**Modify entrypoint.sh to load a custom script when detected**~

    This is a flexible approach, but I do not recommend it.  It would be very easy to break things (e.g., you could naively run varnishd twice).  This could make troubleshooting very frustrating.  If you're modifying/overriding the entrypoint.sh, maybe option 1 or option 3 would be less confusing.

5. ~**Add a trigger for running the varnishadm**~

    Again, I don't recommend this because it is more or less redundant with option 3.  Also, compared to options 1-3, I think it can be a bit confusing to trace where the settings are coming from.


# Minimizing configuration in entrypoint.sh

Overall, minimizing the amount of configuration and moving files in entrypoint.sh can make things a little bit easier to understand, and thus easier to override configurations.

**Comment by chrisjosephs** [comment]
Option 1 would be nice, but for some reason azure app won't automount attached blob storage if you use option 1, and it resets the access key back to 4 digits.  I don't know why this is.  But then if you try to mount the CIFS volume manually, even with priviledged setting passed into the container, it still says the container is not able to mount the CIFS blob storage.  Otherwise I would use option 1.

**Comment by JCFerrerG** [comment]
Hmm, I haven't tried this with Blob Storage, although I do have some experience with Bring-your-own storage and Azure Files...

* Are you using [Blobfuse](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-how-to-mount-container-linux) or something else to mount Blob storage?
* Have you tried it with [Bring-your-own storage in App Service](https://blogs.msdn.microsoft.com/appserviceteam/2018/09/24/announcing-bring-your-own-storage-to-app-service/)?

**Comment by chrisjosephs** [comment]
Using "mount -t cifs //bx1d1drupalwebstordev.file.core.windows.net/drupalfiles /tmp/df -o vers=3.0,username=bx1d1drupalwebstordev,password=PASSWORD,dir_mode=0777,file_mode=0777,sec=ntlmssp"
Which is the mount command provided by azure....   It currently says "permission denied", even if I add "priviledged" to the docker compose in Azure.  However on my local priviledged switch fixes it

Blobfuse is not part of the container currently, so would need to be added to the container to have access to that permanently on startup without forking the image or overriding entrypoint.sh to re-add it every time...

**Comment by LeonZhang77** [comment]
hi, Chris and JCFerrerG...

I just published appsvcorg/drupal-nginx-fpm:0.6, please have a try.

With this version, you can modify files under /home/etc/nginx and /home/etc/varnish, these files will be persisted after restart.

Just in case if you like to restore to default configuration, please check /etc/nginx-bak and /etc/varnish-bak.

Please let me know is it fit your request? 

thanks,
-Leon
